keras_nlp/layers/fnet_encoder.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""FNet encoder block implementation based on `keras.layers.Layer`."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+class FNetEncoder(keras.layers.Layer):
+    """FNet encoder.
+
+    This class follows the architecture of FNet encoder layer in paper
+    "FNet: Mixing Tokens with Fourier Transforms"
+    (https://arxiv.org/abs/2105.03824). Users can instantiate multiple instances
+    of this class to stack up the encoder.
+
+    Args:
+        intermediate_dim: int, defaults to 3072. The hidden size of feedforward
+            network.
+        dropout: float, defaults to 0.1. The dropout value, applied in the
+            feedforward network.
+        activation: string or `tf.keras.activations`, defaults to "gelu". The
+            activation function of feedforward network.
+        layer_norm_epsilon: float, defaults to 1e-12. The epsilon value in layer
+            normalization components.
+        kernel_initializer: tf.keras.initializers initializer, defaults to
+            "glorot_uniform". The kernel initializer for the dense layers.
+        bias_initializer: tf.keras.initializers initializer, defaults to
+            "zeros". The bias initializer for the dense layers.
+        name: string, defaults to None. The name of the layer.
+        **kwargs: other keyword arguments.
+
+    Examples:
+
+    ```python
+    # Create a single FNet encoder layer.
+    encoder = keras_nlp.layers.FNetEncoder(
+        intermediate_dim=64)
+
+    # Create a simple model containing the encoder.
+    input = tf.keras.Input(shape=[4, 6])
+    output = encoder(input)
+    model = tf.keras.Model(inputs=input, outputs=output)
+
+    # Call encoder on the inputs.
+    input_data = tf.random.uniform(shape=[1, 10, 64])
+    output = model(input_data)
+    ```
+
+    References:
+        [Lee-Thorp et al., 2021](https://arxiv.org/abs/2105.03824)
+    """
+
+    def __init__(
+        self,
+        intermediate_dim=3072,
+        dropout=0.1,
```

Comment:
Should default to 0

Reply:
Right, but we are not shipping a model architecture here, we are shipping a reusable layer. I would be inclined to use the same defaults as the TransformerEncoder block for most of the args here so this is essentially a swap in replacement. Having each encoder block we ship use different defaults everywhere would be a lot to track, and lead to bad usability.

Reply:
Right. I'll set the defaults to the ones used by `TransformerEncoder` 👍🏼